crypto/s2n_ecc_evp.c

```diff
@@ -0,0 +1,281 @@
+    S2N_ERROR_IF(ecc_evp_params->evp_pkey == NULL, S2N_ERR_ECDHE_GEN_KEY);
+    return 0;
+}
+
+int s2n_ecc_evp_compute_shared_secret_from_params(struct s2n_ecc_evp_params *private_ecc_evp_params,
+                                                  struct s2n_ecc_evp_params *public_ecc_evp_params,
+                                                  struct s2n_blob *shared_key) {
+    notnull_check(private_ecc_evp_params->negotiated_curve);
+    notnull_check(private_ecc_evp_params->evp_pkey);
+    notnull_check(public_ecc_evp_params->negotiated_curve);
+    notnull_check(public_ecc_evp_params->evp_pkey);
+    S2N_ERROR_IF(private_ecc_evp_params->negotiated_curve->iana_id != public_ecc_evp_params->negotiated_curve->iana_id,
+                 S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
+    GUARD(s2n_ecc_evp_compute_shared_secret(private_ecc_evp_params->evp_pkey, public_ecc_evp_params->evp_pkey,
+                                            shared_key));
+    return 0;
+}
+
+#if (!MODERN_EC_SUPPORTED)
+static int s2n_ecc_evp_calculate_point_length(const EC_POINT *point, const EC_GROUP *group, uint8_t *length) {
+    size_t ret = EC_POINT_point2oct(group, point, POINT_CONVERSION_UNCOMPRESSED, NULL, 0, NULL);
+    S2N_ERROR_IF(ret == 0, S2N_ERR_ECDHE_SERIALIZING);
+    S2N_ERROR_IF(ret > UINT8_MAX, S2N_ERR_ECDHE_SERIALIZING);
+    *length = (uint8_t)ret;
+    return 0;
+}
+
+static int s2n_ecc_evp_write_point_data_snug(const EC_POINT *point, const EC_GROUP *group, struct s2n_blob *out) {
+    size_t ret = EC_POINT_point2oct(group, point, POINT_CONVERSION_UNCOMPRESSED, out->data, out->size, NULL);
+    S2N_ERROR_IF(ret != out->size, S2N_ERR_ECDHE_SERIALIZING);
+    return 0;
+}
+
+static EC_POINT *s2n_ecc_evp_blob_to_point(struct s2n_blob *blob, const EC_KEY *ec_key) {
+    const EC_GROUP *group = EC_KEY_get0_group(ec_key);
+    EC_POINT *point = EC_POINT_new(group);
+    if (point == NULL) {
+        S2N_ERROR_PTR(S2N_ERR_ECDHE_UNSUPPORTED_CURVE);
+    }
+    if (EC_POINT_oct2point(group, point, blob->data, blob->size, NULL) != 1) {
+        EC_POINT_free(point);
+        S2N_ERROR_PTR(S2N_ERR_BAD_MESSAGE);
+    }
+    return point;
+}
+#endif
+
+int s2n_ecc_evp_read_params_point(struct s2n_stuffer *in, int point_size, struct s2n_blob *point_blob) {
+    notnull_check(in);
+    notnull_check(point_blob);
+    gte_check(point_size, 0);
+
+    /* Extract point from stuffer */
+    point_blob->size = point_size;
+    point_blob->data = (uint8_t *)s2n_stuffer_raw_read(in, point_size);
+    notnull_check(point_blob->data);
+
+    return 0;
+}
+
+int s2n_ecc_evp_write_params_point(struct s2n_ecc_evp_params *ecc_evp_params, struct s2n_stuffer *out) {
+    notnull_check(ecc_evp_params);
+    notnull_check(ecc_evp_params->negotiated_curve);
+    notnull_check(ecc_evp_params->evp_pkey);
+    notnull_check(out);
+
+#if MODERN_EC_SUPPORTED
+    out->blob.size = EVP_PKEY_get1_tls_encodedpoint(ecc_evp_params->evp_pkey, &out->blob.data);
```

Comment:
I think you should use a temp blob here instead of using `out->blob`, write the temp blob to the out stuffer, then call `OPENSSL_free()` on the temp blob once it's been written. That way the API for this function is safer, and users don't need to call `OPENSSL_free(out->blob.data)` after the function is called.

Reply:
Will make the change, thank-you.